vars/yamlSubstituteVariables.groovy

```diff
@@ -0,0 +1,314 @@
+            return
+        }
+
+        def manifestData = null;
+        try {
+            // may return a List<Object>  (in case more YAML segments are in the file)
+            // or a Map<String, Object> in case there is just one segment.
+            manifestData = readYaml file: manifestFilePath
+            echo "[YamlSubstituteVariables] Loaded manifest at ${manifestFilePath}!"
+        }
+        catch(Exception ex) {
+            debug("Exception: ${ex}")
+            error "[YamlSubstituteVariables] Could not load manifest file at ${manifestFilePath}."
+            error "[YamlSubstituteVariables] Exception was: ${ex}"
+        }
+
+        def variablesData = null
+        try {
+            // may return a List<Object>  (in case more YAML segments are in the file)
+            // or a Map<String, Object> in case there is just one segment.
+            variablesData = readYaml file: variablesFilePath
+            echo "[YamlSubstituteVariables] Loaded variables file at ${variablesFilePath}!"
+        }
+        catch(Exception ex) {
+            debug("Exception: ${ex}")
+            error "[YamlSubstituteVariables] Could not load manifest variables file at ${variablesFilePath}."
+            error "[YamlSubstituteVariables] Exception was: ${ex}"
+        }
+
+        // substitute all variables.
+        ExecutionContext context = new ExecutionContext()
+        def result = substitute(manifestData, variablesData, context)
+
+        if (context.noVariablesReplaced) {
+            echo "[YamlSubstituteVariables] No variables were found or could be replaced in ${manifestFilePath}. Skipping variable substitution."
+            return
+        }
+
+        // writeYaml won't overwrite the file. You need to delete it first.
+        deleteOutputFileIfPresent(outputFilePath, skipDeletion)
+
+        writeYaml file: outputFilePath, data: result
+
+        echo "[YamlSubstituteVariables] Replaced variables in ${manifestFilePath} with variables from ${variablesFilePath}."
+        echo "[YamlSubstituteVariables] Wrote output file (with variables replaced) at ${outputFilePath}."
+
+        debug("Loaded Manifest: ${manifestData}")
+        debug("Loaded Variables: ${variablesData}")
+        debug("Result: ${result}")
+    }
+}
+
+/**
+ * Removes the given output file, if it exists.
+ * The {@code writeYaml} command won't delete files, so you need to
+ * remove the file yourself, in case it already exists.
+ * @param outputFilePath the path to the output file to write to.
+ */
+private void deleteOutputFileIfPresent(String outputFilePath, Boolean skipDeletion) {
+    File originalFile = new File(outputFilePath)
+    if (originalFile.exists() && !skipDeletion) {
+        boolean deleted = originalFile.delete()
+        echo "[YamlSubstituteVariables] Deleted output file ${outputFilePath}: ${deleted}"
```

Comment:
Not sure if the boolean flag at the end is understandable. Also not sure if 'output file' is understandable. Maybe something like 'Original file ${fileName} has been deleted.' (Must occur only in case the file has been deleted, of course).

Reply:
To make understandable what "output file" actually is, I added the path to the file - so it should be quite clear what got deleted.
The problem is that "original file" would also not be correct, as it really is the file that was specified as "output". In case it was not specified it will default to the original input file. In other cases it will just be another output file (that may need to be deleted to overwrite it).

Reply:
There are basically two things. The first is the name 'output' file. I guess a trouble shooter might not have the context of an output file, it is basically a file which gets deletd, but ok.

The second thing is the boolean flag at the end. I guess it is not that good to understand that the value of the boolean flag indicates that it has been in fact deleted. Just a sentence in the log "File 'xy' deleted" is IMO better than "File 'xy' deleted: true". Of course the sencence only appear in case it has been deleted.

Reply:
:) Ok, I see your point. To make it 100% clear, I will try to get rid of the boolean flag and put it in prose. Not sure about the "output" part though.

Reply:
Has been resolved with a message that clearly indicates that the file was deleted.
Also removed the "output" from the message.